File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Python/9dd17377edc379115587e50dcf6a8a09.py
import sys
from typing import List

class TaskE:
    def solve(self, r: int, c: int, n: int, k: int, x: List[int], y: List[int]) -> int:
        x = [xi - 1 for xi in x]
        y = [yi - 1 for yi in y]
        yx = sorted(zip(y, x))
        head_position = [Position(-1, -1) for _ in range(k)]
        tail_position = [Position(r, -2) for _ in range(k)]
        for i in range(1, k):
            head_position[i].last = head_position[i - 1]
            head_position[i - 1].next = head_position[i]
        for i in range(1, k):
            tail_position[i].last = tail_position[i - 1]
            tail_position[i - 1].next = tail_position[i]
        positions = [Position(xi, i) for i, xi in enumerate(x)]
        answer = 0
        at = 0
        for i in range(c):
            while at < n and yx[at][0] < i:
                at += 1
            for j in range(k - 1):
                head_position[j + 1].link = tail_position[j]
            linked_set = LinkedSet()
            linked_set.add(head_position[k - 1])
            linked_set.add(tail_position[0])
            head_position[k - 1].next = tail_position[0]
            tail_position[0].last = head_position[k - 1]
            delta = 0
            for j in range(i, c):
                while at < n and yx[at][0] == j:
                    where = linked_set.lower(positions[yx[at][1]])
                    current = where
                    for l in range(1, k):
                        delta -= current.value(r)
                        current = current.last
                    next_pos = where.next
                    delta -= next_pos.value(r)
                    current = positions[yx[at][1]]
                    where.next = current
                    current.last = where
                    next_pos.last = current
                    current.next = next_pos
                    current.link = current if k == 1 else where.link
                    linked_set.add(current)
                    delta += next_pos.value(r)
                    delta += current.value(r)
                    current = where
                    for l in range(1, k):
                        current.link = current.link.last
                        delta += current.value(r)
                        current = current.last
                    at += 1
                answer += delta
        return answer

class Position:
    def __init__(self, row: int, id: int):
        self.row = row
        self.id = id
        self.last = None
        self.next = None
        self.link = None
    
    def __lt__(self, other):
        if self.row != other.row:
            return self.row < other.row
        return self.id < other.id
    
    def value(self, r: int) -> int:
        if self.link is None:
            return 0
        return (self.row - self.last.row) * (r - self.link.row)

class LinkedSet:
    def __init__(self):
        self.head = Position(-1, -1)
        self.tail = Position(-2, -2)
        self.head.next = self.tail
        self.tail.last = self.head
    
    def add(self, position: Position):
        position.last = self.tail.last
        position.next = self.tail
        self.tail.last.next = position
        self.tail.last = position
    
    def lower(self, position: Position) -> Position:
        current = self.head.next
        while current != self.tail and current < position:
            current = current.next
        return current

def main():
    r, c, n, k = map(int, sys.stdin.readline().split())
    x = [0] * n
    y = [0] * n
    for i in range(n):
        x[i], y[i] = map(int, sys.stdin.readline().split())
    solver = TaskE()
    answer = solver.solve(r, c, n, k, x, y)
    sys.stdout.write(str(answer))

if __name__ == "__main__":
    main()

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Go/9dd17377edc379115587e50dcf6a8a09.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
)

type Position struct {
	row  int
	id   int
	last *Position
	next *Position
	link *Position
}

type PositionList []*Position

func (p PositionList) Len() int           { return len(p) }
func (p PositionList) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
func (p PositionList) Less(i, j int) bool { return p[i].row < p[j].row || (p[i].row == p[j].row && p[i].id < p[j].id) }

type TaskE struct{}

func (t *TaskE) solve(testNumber int, in *InputReader, out *OutputWriter) {
	r := in.readInt()
	c := in.readInt()
	n := in.readInt()
	k := in.readInt()
	x := make([]int, n)
	y := make([]int, n)
	in.readIntArrays(&x, &y)
	sort.Ints(x)
	sort.Ints(y)
	headPosition := make([]*Position, k)
	for i := 0; i < k; i++ {
		headPosition[i] = &Position{-1, -1, nil, nil, nil}
		if i != 0 {
			headPosition[i].last = headPosition[i-1]
			headPosition[i-1].next = headPosition[i]
		}
	}
	tailPosition := make([]*Position, k)
	for i := 0; i < k; i++ {
		tailPosition[i] = &Position{r, -2, nil, nil, nil}
		if i != 0 {
			tailPosition[i].last = tailPosition[i-1]
			tailPosition[i-1].next = tailPosition[i]
		}
	}
	positions := make([]*Position, n)
	for i := 0; i < n; i++ {
		positions[i] = &Position{x[i], i, nil, nil, nil}
	}
	answer := int64(0)
	for i := 0; i < c; i++ {
		at := 0
		for at < n && y[at] < i {
			at++
		}
		for j := 0; j < k-1; j++ {
			headPosition[j+1].link = tailPosition[j]
		}
		set := make(PositionList, 0)
		set = append(set, headPosition[k-1])
		set = append(set, tailPosition[0])
		headPosition[k-1].next = tailPosition[0]
		tailPosition[0].last = headPosition[k-1]
		delta := int64(0)
		for j := i; j < c; j++ {
			for at < n && y[at] == j {
				where := set.lowerBound(positions[at])
				current := where
				for l := 1; l < k; l++ {
					delta -= current.value(r)
					current = current.last
				}
				next := where.next
				delta -= next.value(r)
				current = positions[at]
				where.next = current
				current.last = where
				next.last = current
				current.next = next
				if k == 1 {
					current.link = current
				} else {
					current.link = where.link
				}
				set = append(set, current)
				delta += next.value(r)
				delta += current.value(r)
				current = where
				for l := 1; l < k; l++ {
					current.link = current.link.last
					delta += current.value(r)
					current = current.last
				}
				at++
			}
			answer += delta
		}
	}
	out.printLine(answer)
}

func (p *Position) value(r int) int64 {
	if p.link == nil {
		return 0
	}
	return int64((p.row - p.last.row) * (r - p.link.row))
}

type IntIterator interface {
	value() int
	advance() bool
	isValid() bool
}

type IntComparator interface {
	compare(first, second int) int
}

type IntCollection interface {
	IntStream
	size() int
	add(value int)
	addAll(values IntStream)
}

type IntList interface {
	IntCollection
	get(index int) int
	set(index, value int)
	addAt(index, value int)
	removeAt(index int)
	sort(comparator IntComparator) IntList
	subList(from, to int) IntList
}

type IntArray struct {
	data []int
}

func (a *IntArray) size() int {
	return len(a.data)
}

func (a *IntArray) get(index int) int {
	return a.data[index]
}

func (a *IntArray) set(index, value int) {
	a.data[index] = value
}

func (a *IntArray) addAt(index, value int) {
	a.data = append(a.data[:index], append([]int{value}, a.data[index:]...)...)
}

func (a *IntArray) removeAt(index int) {
	a.data = append(a.data[:index], a.data[index+1:]...)
}

func (a *IntArray) add(value int) {
	a.data = append(a.data, value)
}

func (a *IntArray) addAll(values IntStream) {
	for it := values.intIterator(); it.isValid(); it.advance() {
		a.add(it.value())
	}
}

func (a *IntArray) sort(comparator IntComparator) IntList {
	sort.Slice(a.data, func(i, j int) bool {
		return comparator.compare(a.data[i], a.data[j]) < 0
	})
	return a
}

func (a *IntArray) subList(from, to int) IntList {
	return &IntArray{data: a.data[from:to]}
}

type IntStream interface {
	IntIterator
	iterator() Iterator
	compareTo(c IntStream) int
}

type InputReader struct {
	reader *bufio.Reader
}

func NewInputReader(r io.Reader) *InputReader {
	return &InputReader{reader: bufio.NewReader(r)}
}

func (r *InputReader) readInt() int {
	var sign, num int = 1, 0
	var err error
	for {
		var b byte
		b, err = r.reader.ReadByte()
		if err != nil {
			break
		}
		if b >= '0' && b <= '9' {
			num = int(b - '0')
			break
		}
		if b == '-' {
			sign = -1
			break
		}
	}
	for {
		var b byte
		b, err = r.reader.ReadByte()
		if err != nil || b < '0' || b > '9' {
			break
		}
		num = num*10 + int(b-'0')
	}
	return num * sign
}

func (r *InputReader) readIntArrays(arrays ...*[]int) {
	for i := 0; i < len((*arrays)[0]); i++ {
		for j := 0; j < len(arrays); j++ {
			(*arrays)[j][i] = r.readInt()
		}
	}
}

type OutputWriter struct {
	writer *bufio.Writer
}

func NewOutputWriter(w io.Writer) *OutputWriter {
	return &OutputWriter{writer: bufio.NewWriter(w)}
}

func (w *OutputWriter) printLine(i int64) {
	fmt.Fprintln(w.writer, i)
}

func (w *OutputWriter) close() {
	w.writer.Flush()
}

func main() {
	input := NewInputReader(os.Stdin)
	output := NewOutputWriter(os.Stdout)
	defer output.close()
	taskE := &TaskE{}
	taskE.solve(1, input, output)
}
